compiler/rustc_borrowck/src/borrow_set.rs

```diff
@@ -253,19 +253,61 @@ impl<'a, 'tcx> Visitor<'tcx> for GatherBorrows<'a, 'tcx> {
             }
 
             let region = region.as_var();
-
-            let borrow = BorrowData {
+            let borrow = |activation_location| BorrowData {
                 kind,
                 region,
                 reserve_location: location,
-                activation_location: TwoPhaseActivation::NotTwoPhase,
+                activation_location,
                 borrowed_place,
                 assigned_place: *assigned_place,
             };
-            let (idx, _) = self.location_map.insert_full(location, borrow);
-            let idx = BorrowIndex::from(idx);
 
-            self.insert_as_pending_if_two_phase(location, assigned_place, kind, idx);
```

Comment:
this felt very confusing to read to me. we first insert the borrow with the wrong activation location (we assume its not two phase). and then `insert_as_pending_if_two_phase` does more than it says and actually *fixes up* the activation location to what it should have been from the start

I inlined this function as it only had one call site and deferred inserting into the location map until we actually choose a `TwoPhaseActivation` to use.